kernel/comps/virtio/src/device/input/device.rs

```diff
@@ -181,44 +185,133 @@ impl InputDevice {
         size
     }
 
-    /// Negotiate features for the device specified bits 0~23
-    pub(crate) fn negotiate_features(features: u64) -> u64 {
-        assert_eq!(features, 0);
-        0
-    }
-}
-
-impl aster_input::InputDevice for InputDevice {
-    fn handle_irq(&self) -> Option<()> {
-        // one interrupt may contains serval input, so it should loop
-        loop {
-            let Some(event) = self.pop_pending_event() else {
-                return Some(());
+    fn handle_irq(&self) {
+        // Returns ture if there may be more events to handle
+        let handle_event = |event: &EventBuf| -> bool {
+            let event: VirtioInputEvent = {
+                let mut reader = event.reader();
+                reader.read_val()
             };
+
             match event.event_type {
-                0 => return Some(()),
+                0 => return false,
                 // Keyboard
                 1 => {}
                 // TODO: Support mouse device.
-                _ => continue,
+                _ => return true,
             }
+
             let status = match event.value {
                 1 => KeyStatus::Pressed,
                 0 => KeyStatus::Released,
-                _ => return Some(()),
+                _ => return false,
             };
+
             let event = InputEvent::KeyBoard(Key::try_from(event.code).unwrap(), status);
             info!("Input Event:{:?}", event);
 
             let callbacks = self.callbacks.lock();
```

Comment:
This callback list is mostly read. Protect it with a read-write lock. Apply this optimization to all driver that contains a callback list in a separate commit.